### ISSUE TITLE
false positive `unused-extern-crates`

### ISSUE THREAD

**Issue by matthiaskrgr** [rust-lang/rust#148140]
Using the following flags

`--force-warn unused-extern-crates`

this code:

```rust
// No error, panic handler is supplied by libstd linked though the empty library.

//@ check-pass
//@ aux-build: cfg_false_lib.rs

#![no_std]

extern crate proc_macro;

fn main() {
    panic!()
}

```

caused the following diagnostics:

```
    Checking _5ba66f8e62a4839c0ab39d6e4a771d7ada6b5b40 v0.1.0 (/tmp/icemaker_global_tempdir.ZzRBF4urpxex/icemaker_clippyfix_tempdir.aeQgi7GJI5WY/_5ba66f8e62a4839c0ab39d6e4a771d7ada6b5b40)
warning: unused extern crate
 --> src/main.rs:8:1
  |
8 | extern crate proc_macro;
  | ^^^^^^^^^^^^^^^^^^^^^^^^ unused
  |
  = note: requested on the command line with `--force-warn unused-extern-crates`
help: remove the unused `extern crate`
  |
8 - extern crate proc_macro;
  |

warning: `_5ba66f8e62a4839c0ab39d6e4a771d7ada6b5b40` (bin "_5ba66f8e62a4839c0ab39d6e4a771d7ada6b5b40") generated 1 warning (run `cargo clippy --fix --bin "_5ba66f8e62a4839c0ab39d6e4a771d7ada6b5b40" -p _5ba66f8e62a4839c0ab39d6e4a771d7ada6b5b40` to apply 1 suggestion)
    Finished `dev` profile [unoptimized + debuginfo] target(s) in 0.21s

```

However after applying these diagnostics, the resulting code:

```rust
// No error, panic handler is supplied by libstd linked though the empty library.

//@ check-pass
//@ aux-build: cfg_false_lib.rs

#![no_std]


fn main() {
    panic!()
}

```
no longer compiled:

```
    Checking _5ba66f8e62a4839c0ab39d6e4a771d7ada6b5b40 v0.1.0 (/tmp/icemaker_global_tempdir.ZzRBF4urpxex/icemaker_clippyfix_tempdir.aeQgi7GJI5WY/_5ba66f8e62a4839c0ab39d6e4a771d7ada6b5b40)
error: `#[panic_handler]` function required, but not found

error: unwinding panics are not supported without std
  |
  = help: using nightly cargo, use -Zbuild-std with panic="abort" to avoid unwinding
  = note: since the core library is usually precompiled with panic="unwind", rebuilding your crate with panic="abort" may not be enough to fix the problem

error: could not compile `_5ba66f8e62a4839c0ab39d6e4a771d7ada6b5b40` (bin "_5ba66f8e62a4839c0ab39d6e4a771d7ada6b5b40") due to 2 previous errors
warning: build failed, waiting for other jobs to finish...

```

Version:
```
rustc 1.93.0-nightly (a8664a153 2025-10-26)
binary: rustc
commit-hash: a8664a1534913ccff491937ec2dc7ec5d973c2bd
commit-date: 2025-10-26
host: x86_64-unknown-linux-gnu
release: 1.93.0-nightly
LLVM version: 21.1.3

```

**Comment by chenyukang** [comment]
I looked into this a bit and I’m unsure what level of fix would be preferred here.

So I’m wondering whether the desired behavior should be:
- A narrow fix is to special-case `extern crate proc_macro` for `#![no_std]` crates that emit a linked artifact, since `proc_macro` depends on `std`. That fixes the reported case, but it feels a bit ad-hoc.
- Keep the lint warning, but avoid emitting a machine-applicable removal suggestion when `std` is absent from the extern prelude and the crate emits a linked artifact.
- Try to determine more precisely whether the extern crate is the dependency that pulls in std/panic runtime, though that seems too heavy for this lint. I think it does not worth it.

What's your opinion on this? @petrochenkov

**Comment by petrochenkov** [comment]
Not worth fixing, IMO.
If the set of available and required lang items was known, we could do better, but it would require either delaying the lint, or doing the lang item collection earlier, i.e. not worth it.
We could use "no_std" as an approximation of "requires lang items", but it would introduce false negatives instead.

`unused_extern_crates` is known to have other uncheckable false positives too, e.g. when the crate in question provides some necessary link time symbol.
The lint is easy to suppress by doing `extern crate foo as _;` (preferably with a comment).
